File: dummy/java/LibraryManagementSystem.java
import java.io.*;
import java.util.*;
import java.time.LocalDate;
import java.time.temporal.ChronoUnit;

class Book {
    private String title;
    private String author;
    private String isbn;
    private boolean isCheckedOut;
    private LocalDate dueDate;

    public Book(String title, String author, String isbn) {
        this.title = title;
        this.author = author;
        this.isbn = isbn;
        this.isCheckedOut = false;
        this.dueDate = null;
    }

    // Getters and setters
    public String getTitle() { return title; }
    public String getAuthor() { return author; }
    public String getIsbn() { return isbn; }
    public boolean isCheckedOut() { return isCheckedOut; }
    public LocalDate getDueDate() { return dueDate; }

    public void checkOut() {
        isCheckedOut = true;
        dueDate = LocalDate.now().plusDays(14); // Due in 14 days
    }

    public void returnBook() {
        isCheckedOut = false;
        dueDate = null;
    }

    @Override
    public String toString() {
        return String.format("Title: %s, Author: %s, ISBN: %s, Checked Out: %s, Due Date: %s",
                title, author, isbn, isCheckedOut, dueDate);
    }
}

class Library {
    private List<Book> books;
    private static final String LIBRARY_FILE = "library.txt";

    public Library() {
        books = new ArrayList<>();
        loadLibrary();
    }

    public void addBook(Book book) {
        books.add(book);
    }

    public void removeBook(String isbn) {
        books.removeIf(book -> book.getIsbn().equals(isbn));
    }

    public Book findBook(String isbn) {
        return books.stream()
                .filter(book -> book.getIsbn().equals(isbn))
                .findFirst()
                .orElse(null);
    }

    public void checkOutBook(String isbn) {
        Book book = findBook(isbn);
        if (book != null && !book.isCheckedOut()) {
            book.checkOut();
            System.out.println("Book checked out successfully.");
        } else {
            System.out.println("Book not available for checkout.");
        }
    }

    public void returnBook(String isbn) {
        Book book = findBook(isbn);
        if (book != null && book.isCheckedOut()) {
            book.returnBook();
            System.out.println("Book returned successfully.");
        } else {
            System.out.println("Invalid return operation.");
        }
    }

    public void displayOverdueBooks() {
        LocalDate today = LocalDate.now();
        books.stream()
                .filter(Book::isCheckedOut)
                .filter(book -> book.getDueDate().isBefore(today))
                .forEach(book -> System.out.println(book + " - Days Overdue: " +
                        ChronoUnit.DAYS.between(book.getDueDate(), today)));
    }

    public void saveLibrary() {
        try (PrintWriter writer = new PrintWriter(new FileWriter(LIBRARY_FILE))) {
            for (Book book : books) {
                writer.println(String.format("%s,%s,%s,%b,%s",
                        book.getTitle(), book.getAuthor(), book.getIsbn(),
                        book.isCheckedOut(), book.getDueDate()));
            }
            System.out.println("Library saved successfully.");
        } catch (IOException e) {
            System.err.println("Error saving library: " + e.getMessage());
        }
    }

    private void loadLibrary() {
        try (BufferedReader reader = new BufferedReader(new FileReader(LIBRARY_FILE))) {
            String line;
            while ((line = reader.readLine()) != null) {
                String[] parts = line.split(",");
                Book book = new Book(parts[0], parts[1], parts[2]);
                if (Boolean.parseBoolean(parts[3])) {
                    book.checkOut();
                    book.getDueDate().parse(parts[4]);
                }
                books.add(book);
            }
            System.out.println("Library loaded successfully.");
        } catch (IOException e) {
            System.err.println("Error loading library: " + e.getMessage());
        }
    }
}

public class LibraryManagementSystem {
    private static Library library = new Library();
    private static Scanner scanner = new Scanner(System.in);

    public static void main(String[] args) {
        while (true) {
            displayMenu();
            int choice = getIntInput("Enter your choice: ");
            switch (choice) {
                case 1: addBook(); break;
                case 2: removeBook(); break;
                case 3: findBook(); break;
                case 4: checkOutBook(); break;
                case 5: returnBook(); break;
                case 6: displayOverdueBooks(); break;
                case 7: library.saveLibrary(); break;
                case 8: System.out.println("Exiting..."); return;
                default: System.out.println("Invalid choice. Please try again.");
            }
        }
    }

    private static void displayMenu() {
        System.out.println("\n--- Library Management System ---");
        System.out.println("1. Add a book");
        System.out.println("2. Remove a book");
        System.out.println("3. Find a book");
        System.out.println("4. Check out a book");
        System.out.println("5. Return a book");
        System.out.println("6. Display overdue books");
        System.out.println("7. Save library");
        System.out.println("8. Exit");
    }

    private static void addBook() {
        String title = getStringInput("Enter book title: ");
        String author = getStringInput("Enter book author: ");
        String isbn = getStringInput("Enter book ISBN: ");
        library.addBook(new Book(title, author, isbn));
        System.out.println("Book added successfully.");
    }

    private static void removeBook() {
        String isbn = getStringInput("Enter ISBN of book to remove: ");
        library.removeBook(isbn);
        System.out.println("Book removed successfully.");
    }

    private static void findBook() {
        String isbn = getStringInput("Enter ISBN of book to find: ");
        Book book = library.findBook(isbn);
        if (book != null) {
            System.out.println("Book found: " + book);
        } else {
            System.out.println("Book not found.");
        }
    }

    private static void checkOutBook() {
        String isbn = getStringInput("Enter ISBN of book to check out: ");
        library.checkOutBook(isbn);
    }

    private static void returnBook() {
        String isbn = getStringInput("Enter ISBN of book to return: ");
        library.returnBook(isbn);
    }

    private static void displayOverdueBooks() {
        library.displayOverdueBooks();
    }

    private static String getStringInput(String prompt) {
        System.out.print(prompt);
        return scanner.nextLine();
    }

    private static int getIntInput(String prompt) {
        while (true) {
            try {
                System.out.print(prompt);
                return Integer.parseInt(scanner.nextLine());
            } catch (NumberFormatException e) {
                System.out.println("Invalid input. Please enter a number.");
            }
        }
    }
}

File: dummy/python/LibraryManagementSystem.py
import datetime
import os
from dataclasses import dataclass
from typing import List, Optional


@dataclass
class Book:
    title: str
    author: str
    isbn: str
    is_checked_out: bool = False
    due_date: Optional[datetime.date] = None

    def check_out(self) -> None:
        self.is_checked_out = True
        self.due_date = datetime.date.today() + datetime.timedelta(days=14)

    def return_book(self) -> None:
        self.is_checked_out = False
        self.due_date = None

    def __str__(self) -> str:
        return (
            f"Title: {self.title}, Author: {self.author}, ISBN: {self.isbn}, "
            f"Checked Out: {self.is_checked_out}, Due Date: {self.due_date}"
        )


class Library:
    LIBRARY_FILE = "library.txt"

    def __init__(self) -> None:
        self.books: List[Book] = []
        self.load_library()

    def add_book(self, book: Book) -> None:
        self.books.append(book)

    def remove_book(self, isbn: str) -> None:
        self.books = [book for book in self.books if book.isbn != isbn]

    def find_book(self, isbn: str) -> Optional[Book]:
        return next((book for book in self.books if book.isbn == isbn), None)

    def check_out_book(self, isbn: str) -> None:
        book = self.find_book(isbn)
        if book and not book.is_checked_out:
            book.check_out()
            print("Book checked out successfully.")
        else:
            print("Book not available for checkout.")

    def return_book(self, isbn: str) -> None:
        book = self.find_book(isbn)
        if book and book.is_checked_out:
            book.return_book()
            print("Book returned successfully.")
        else:
            print("Invalid return operation.")

    def display_overdue_books(self) -> None:
        today = datetime.date.today()
        for book in self.books:
            if book.is_checked_out and book.due_date and book.due_date < today:
                overdue_days = (today - book.due_date).days
                print(f"{book} - Days Overdue: {overdue_days}")

    def save_library(self) -> None:
        try:
            with open(self.LIBRARY_FILE, "w") as file:
                for book in self.books:
                    file.write(
                        f"{book.title},{book.author},{book.isbn},"
                        f"{book.is_checked_out},{book.due_date}\n"
                    )
            print("Library saved successfully.")
        except IOError as e:
            print(f"Error saving library: {e}")

    def load_library(self) -> None:
        if not os.path.exists(self.LIBRARY_FILE):
            return

        try:
            with open(self.LIBRARY_FILE, "r") as file:
                for line in file:
                    parts = line.strip().split(",")
                    if len(parts) < 4:
                        continue
                    book = Book(parts[0], parts[1], parts[2])
                    if parts[3] == "True":
                        book.check_out()
                        if len(parts) > 4 and parts[4] != "None":
                            book.due_date = datetime.date.fromisoformat(parts[4])
                    self.books.append(book)
            print("Library loaded successfully.")
        except IOError as e:
            print(f"Error loading library: {e}")


def get_string_input(prompt: str) -> str:
    return input(prompt)


def get_int_input(prompt: str) -> int:
    while True:
        try:
            return int(input(prompt))
        except ValueError:
            print("Invalid input. Please enter a number.")


def display_menu() -> None:
    print("\n--- Library Management System ---")
    print("1. Add a book")
    print("2. Remove a book")
    print("3. Find a book")
    print("4. Check out a book")
    print("5. Return a book")
    print("6. Display overdue books")
    print("7. Save library")
    print("8. Exit")


def main() -> None:
    library = Library()
    while True:
        display_menu()
        choice = get_int_input("Enter your choice: ")
        if choice == 1:
            title = get_string_input("Enter book title: ")
            author = get_string_input("Enter book author: ")
            isbn = get_string_input("Enter book ISBN: ")
            library.add_book(Book(title, author, isbn))
            print("Book added successfully.")
        elif choice == 2:
            isbn = get_string_input("Enter ISBN of book to remove: ")
            library.remove_book(isbn)
            print("Book removed successfully.")
        elif choice == 3:
            isbn = get_string_input("Enter ISBN of book to find: ")
            book = library.find_book(isbn)
            print("Book found: " + str(book) if book else "Book not found.")
        elif choice == 4:
            isbn = get_string_input("Enter ISBN of book to check out: ")
            library.check_out_book(isbn)
        elif choice == 5:
            isbn = get_string_input("Enter ISBN of book to return: ")
            library.return_book(isbn)
        elif choice == 6:
            library.display_overdue_books()
        elif choice == 7:
            library.save_library()
        elif choice == 8:
            print("Exiting...")
            break
        else:
            print("Invalid choice. Please try again.")


if __name__ == "__main__":
    main()
